### PR TITLE
Fixes #23031 - parse server duid's with nested double quotes

### PR DESCRIPTION
**Pull request by dmitri-d**

### CODE REVIEW

**Review by ekohl — APPROVED** [review]

File: modules/dhcp_common/isc/configuration_parser.rb [hunk 1]
```diff
@@ -209,10 +209,11 @@ def server_duid
           Rsec::Fail.reset
           keyword = word('server-duid').fail 'keyword_server_duid'
           anything = /[^;,{}\s]+/.r
+          literal = /"([^"]|\")*"/.r
```
Comment (lzap) [inline modules/dhcp_common/isc/configuration_parser.rb:212]: It looks like ISC configuration parser treats all strings like that, isn't cleaner to use this in `LITERAL` constant?

https://github.com/42wim/isc-dhcp/blob/f54a146c7fe88889d60f0c1aa8e6f04707f95223/common/conflex.c#L512

**Comment by lzap** [comment]
If it won't break our tests, I think we should go with:

    LITERAL = /"([^"]|\")*"/.r

Single quotes do not appear to be supported by the parser, if I am looking into the correct place in the codebase. https://lists.isc.org/pipermail/dhcp-users/2011-February/013135.html

**Comment by dmitri-d** [comment]
I think the ultimate goal  should be parsing strings the same way as dhcpd's parser does. Unfortunately, I don't think it is doable using a non-monstrous regex (if at all). The reason I didn't reuse the regex the way you are suggesting is that it will also match comma-separated strings as one string, which breaks the parser.

DUID is an outlier, as majority of the strings are host or option names and values, I think we are pretty safe with the fix in the PR. I'll add extending rsec parser with a custom string parser as a feature to smart-proxy's backlog.

**Review by lzap — APPROVED** [review]
Ok that's good. Thanks.

**Comment by mark-wagner** [comment]
I think a similar change will be needed for the uid within a lease.

**Comment by lzap** [comment]
I also think doing this for all strings won't hurt, the ISC parser does seem to work that way.

**Comment by alvinstarr** [comment]
I believe that  literal should be  `literal = /"([^"]|\\")*"/.r`
The patch you suggested is in the version of 1.17 that I have installed but I was still getting duid errors until I added the second \.
I am really unfamiliar with ruby so I may well be wrong.

**Comment by mark-wagner** [comment]
@alvinstarr that regexp doesn't work for. Here is the uid that was tripping me up:

`uid "\001X\357h}g\""`

http://rubular.com/r/gqxRXXs1YZ

The regexp in https://stackoverflow.com/questions/249791/regex-for-quoted-string-with-escaping-quotes seems to work but may have other problems.

http://rubular.com/r/vRvdrzFYxe

**Comment by dmitri-d** [comment]
Please see my comment above. Feel free to experiment, but please run tests before proposing a fix. Please also note complicated regular expressions will be rejected.

**Comment by alvinstarr** [comment]
This may be better 
`"([^\\"]|\\.)*"`
I believe the string consists of:
char - a single character
\### - 3 digits 0-7 for octal characters
\char - an escaped character
This of course only works for ASCII characters if my memories of C strings is correct.

**Comment by dmitri-d** [comment]
I am locking this conversation. If you are seeing this issue, please update the redmine ticket referenced in this PR, or create a new one.

**Comment by lzap** [comment]
I am not sure if locking was necessary here, "this might be better" is appropriate comment although I do agree some of them can be useless when not properly tried and tested. However, I think leaving those comments "as-is" is probably preferred than locking discussion.

**Pull request merged**